api/apps/api/test/project/get-export-file.e2e-spec.ts

```diff
@@ -0,0 +1,241 @@
+  ExportId,
+} from '@marxan-api/modules/clone/export/domain';
+import { ProjectChecker } from '@marxan-api/modules/projects/project-checker/project-checker.service';
+import { FileRepository } from '@marxan/files-repository';
+import { FixtureType } from '@marxan/utils/tests/fixture-type';
+import { CommandBus, EventBus, IEvent } from '@nestjs/cqrs';
+import { getRepositoryToken } from '@nestjs/typeorm';
+import { isLeft } from 'fp-ts/lib/These';
+import { Subscription } from 'rxjs';
+import { Readable } from 'stream';
+import * as request from 'supertest';
+import { Connection, Repository } from 'typeorm';
+import { Class } from 'utility-types';
+import { GivenProjectExists } from '../steps/given-project';
+import { GivenUserExists } from '../steps/given-user-exists';
+import { GivenUserIsLoggedIn } from '../steps/given-user-is-logged-in';
+import { bootstrapApplication } from '../utils/api-application';
+import { OrganizationsTestUtils } from '../utils/organizations.test.utils';
+import { ProjectCheckerFake } from '../utils/project-checker.service-fake';
+import { ProjectsTestUtils } from '../utils/projects.test.utils';
+
+async function untilEventIsEmitted<T extends IEvent>(
+  eventClass: Class<T>,
+  eventBus: EventBus,
+): Promise<T> {
+  let subscription: Subscription;
+  const event = await new Promise<T>((resolve) => {
+    subscription = eventBus.subscribe((event) => {
+      if (event instanceof eventClass) {
+        resolve(event);
+      }
+    });
+  });
+
+  subscription!.unsubscribe();
+
+  return event;
+}
+
+let fixtures: FixtureType<typeof getFixtures>;
+
+beforeEach(async () => {
+  fixtures = await getFixtures();
+}, 10_000);
+
+afterEach(async () => {
+  await fixtures?.cleanup();
+});
+
+test('should forbid downloading export file to unrelated users', async () => {
+  await fixtures.GivenProjectWasCreated();
+  await fixtures.GivenExportWasRequested();
+
+  await fixtures.WhenExportFileIsReady();
+  const response = await fixtures.WhenUnrelatedUserRequestExportFile();
+
+  fixtures.ThenForbiddenIsReturned(response);
+});
+
+test('should permit downloading public project export file for unrelated users', async () => {
+  await fixtures.GivenProjectWasCreated();
+  await fixtures.GivenProjectIsPublic();
+  await fixtures.GivenExportWasRequested();
+
+  await fixtures.WhenExportFileIsReady();
+  const response = await fixtures.WhenUnrelatedUserRequestExportFile();
+
+  fixtures.ThenFileIsDownloaded(response);
+});
+
+test('should permit downloading export file for owner users ', async () => {
+  await fixtures.GivenProjectWasCreated();
+  await fixtures.GivenExportWasRequested();
+
+  await fixtures.WhenExportFileIsReady();
+  const response = await fixtures.WhenOwnerUserRequestExportFile();
+
+  fixtures.ThenFileIsDownloaded(response);
+});
+
+test('should permit downloading export file for contributor users ', async () => {
+  await fixtures.GivenProjectWasCreated();
+  await fixtures.GivenContributorWasAddedToProject();
```

Comment:
Minor, but maybe we can simplify this a bit for the checks that involve users with a role on the project - the fact that users with relevant roles have been granted access to the project should be more or less implicit when we check if a user with given role can download the export archive, so I would grant the roles to the users transparently within `GivenProjectWasCreated()`, or keep a separate step for this but grant all roles at once. Though again, I'd avoid exposing the role granting as a `given` step to avoid confusion - at least in my view: doing all this within the project setup step seems to work best, in my view.

Reply:
Double checking this, having a step called `WhenINSERT_ROLEUserRequestExportFile` is more than enought 👌🏽